File: java/src/0445.java
import static org.assertj.core.api.Assertions.assertThat;
import static structs.util.LinkedList.of;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.LinkedList;

import structs.linkedlist.ListNode;

/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        var s1 = new LinkedList<Integer>();
        var s2 = new LinkedList<Integer>();
        for (var p = l1; p != null; p = p.next) {
            s1.push(p.val);
        }
        for (var q = l2; q != null; q = q.next) {
            s2.push(q.val);
        }
        var carry = 0;
        ListNode prev = null;
        while (!s1.isEmpty() || !s2.isEmpty() || carry != 0) {
            if (!s1.isEmpty()) {
                carry += s1.pop();
            }
            if (!s2.isEmpty()) {
                carry += s2.pop();
            }
            var curr = new ListNode(carry % 10, prev);
            carry /= 10;
            prev = curr;
        }
        return prev;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.addTwoNumbers(of(7, 2, 4, 3), of(5, 6, 4)))
                .isEqualTo(of(7, 8, 0, 7));
        assertThat(o.addTwoNumbers(of(2, 4, 3), of(5, 6, 4)))
                .isEqualTo(of(8, 0, 7));
        assertThat(o.addTwoNumbers(of(0), of(0))).isEqualTo(of(0));

        var start = Instant.now();
        for (var i = 0; i < 1000000; i++) {
            o.addTwoNumbers(of(7, 2, 4, 3), of(5, 6, 4));
            o.addTwoNumbers(of(2, 4, 3), of(5, 6, 4));
            o.addTwoNumbers(of(0), of(0));
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: python/src/0445.py
from typing import Optional

from structs.ListNode import ListNode, create

from collections import deque


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        s1, s2 = deque(), deque()
        p, q = l1, l2
        while p:
            s1.append(p.val)
            p = p.next
        while q:
            s2.append(q.val)
            q = q.next
        carry, prev = 0, None
        while s1 or s2 or carry:
            if s1:
                carry += s1.pop()
            if s2:
                carry += s2.pop()
            curr = ListNode(carry % 10, prev)
            carry //= 10
            prev = curr
        return prev


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(7, 2, 4, 3), create(5, 6, 4), create(7, 8, 0, 7)),
        (create(2, 4, 3), create(5, 6, 4), create(8, 0, 7)),
        (create(0), create(0), create(0)),
        (None, None, None),
    )

    class Test(unittest.TestCase):

        def test(self):
            addTwoNumbers = Solution().addTwoNumbers
            for l1, l2, excepted in CASES:
                actual = addTwoNumbers(l1, l2)
                self.assertEqual(actual, excepted)

    unittest.main()
